shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -348,12 +345,42 @@ - (BOOL)createShell:(NSString*)entrypoint libraryURI:(NSString*)libraryURI {
     FML_LOG(ERROR) << "Could not start a shell FlutterEngine with entrypoint: "
                    << entrypoint.UTF8String;
   } else {
+    [self setupChannels];
+    if (!_platformViewsController) {
+      _platformViewsController.reset(new shell::FlutterPlatformViewsController());
+    }
+    _publisher.reset([[FlutterObservatoryPublisher alloc] init]);
     [self maybeSetupPlatformViewChannels];
   }
 
   return _shell != nullptr;
 }
 
+- (void)shutdown {
+  if (_shell == nullptr) {
+    return;
+  }
+  [self setViewController:nil];
+
+  _publisher.reset();
+
+  _platformViewsController.reset();
+
+  _platformPlugin.reset();
+  _textInputPlugin.reset();
+  _localizationChannel.reset();
+  _navigationChannel.reset();
+  _platformChannel.reset();
+  _platformViewsChannel.reset();
+  _textInputChannel.reset();
+  _lifecycleChannel.reset();
+  _systemChannel.reset();
+  _settingsChannel.reset();
+
+  _shell.reset();
+  _threadHost.Reset();
```

Comment:
This seems a bit brittle to me, since anyone adding new members probably will need to remember to update this method.

One alternative would be to wrap all of these members within another `struct`, and then you can just clobber that `struct` with a default-initialized one.

If this were pure C++, I'd even consider doing something like `*this = FlutterEngine();` (and explicitly saving/restoring members we want to preserve), but I don't know how that'd work with Objective-C++.

Reply:
Agreed - I'll clean this up...

Reply:
I've moved most of these to a method with a comment - I also found a couple that weren't necessary in there.  The basic problem is that MessageChannels get the Engine pointer and retain it - and we don't have a nice, non-breaking way to deal with that via a WeakPtr or something because it's part of the public Objective-C API.  I know this approach still isn't quite ideal, but hopefully it's better.